src/store.ts

```diff
@@ -413,32 +417,112 @@ export default function Store(
+    }
+
+    // Passed in config flags take priority over defaults
+    flags[ReportBatching] = config.flags[ReportBatching] || false;
+    // The server returns stringified numbers, sowe need to parse
+    flags[EventBatchingIntervalMillis] =
+        parseNumber(config.flags[EventBatchingIntervalMillis]) ||
+        Constants.DefaultConfig.uploadInterval;
+    flags[OfflineStorage] = config.flags[OfflineStorage] || '0';
+    flags[DirectUrlRouting] = config.flags[DirectUrlRouting] === 'True';
+
+    return flags;
+}
+
+export function processBaseUrls(
+    config: SDKInitConfig,
+    flags: IFeatureFlags,
+    apiKey?: string
+): Dictionary<string> {
+    // an API key is not present in a webview only mode. In this case, no baseUrls are needed
+    if (!apiKey) {
+        return {};
+    }
+
+    // Set default baseUrls
+    let baseUrls: Dictionary<string>;
+
+    // When direct URL routing is false, update baseUrls based custom urls
+    // passed to the config
+    if (flags.directURLRouting) {
+        return processDirectBaseUrls(config, apiKey);
+    } else {
+        return processCustomBaseUrls(config);
     }
 }
+
+function processCustomBaseUrls(config: SDKInitConfig): Dictionary<string> {
+    const defaultBaseUrls: Dictionary<string> = Constants.DefaultBaseUrls;
+    const newBaseUrls: Dictionary<string> = {};
+
+    // If there is no custo base url, we use the default base url
+    for (let baseUrlKey in defaultBaseUrls) {
+        newBaseUrls[baseUrlKey] =
+            config[baseUrlKey] || defaultBaseUrls[baseUrlKey];
+    }
+
+    return newBaseUrls;
+}
+
+function processDirectBaseUrls(
+    config: SDKInitConfig,
+    apiKey: string
+): Dictionary {
+    const defaultBaseUrls = Constants.DefaultBaseUrls;
+    const directBaseUrls: Dictionary<string> = {};
+    // When Direct URL Routing is true, we create a new set of baseUrls that
+    // include the silo in the urls.  mParticle API keys are prefixed with the
+    // silo and a hyphen (ex. "us1-", "us2-", "eu1-").  us1 was the first silo,
+    // and before other silos existed, there were no prefixes and all apiKeys
+    // were us1. As such, if we split on a '-' and the resulting array length
+    // is 1, then it is an older APIkey that should route to us1.
+    // When splitKey.length is greater than 1, then splitKey[0] will be
+    // us1, us2, eu1, au1, or st1, etc as new silos are added
+    const DEFAULT_SILO = 'us1';
+    const splitKey: Array<string> = apiKey.split('-');
+    const routingPrefix: string =
+        splitKey.length <= 1 ? DEFAULT_SILO : splitKey[0];
+
+    for (let baseUrlKey in defaultBaseUrls) {
+        // Any custom endpoints passed to config will take priority over direct
+        // mapping to the silo
+        if (baseUrlKey === 'configUrl') {
+            directBaseUrls[baseUrlKey] =
+                config[baseUrlKey] || defaultBaseUrls[baseUrlKey];
+            continue;
+        }
+
+        if (config.hasOwnProperty(baseUrlKey)) {
```

Comment:
maybe add comment above this line to say that a CNAME overrides direct url routing